File: app/src/main/java/acr/browser/lightning/browser/fragment/TabsFragment.java
package acr.browser.lightning.browser.fragment;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewCompat;
import android.support.v4.widget.TextViewCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.LayoutManager;
import android.support.v7.widget.SimpleItemAnimator;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import javax.inject.Inject;

import acr.browser.lightning.R;
import acr.browser.lightning.browser.TabsManager;
import acr.browser.lightning.BrowserApp;
import acr.browser.lightning.browser.TabsView;
import acr.browser.lightning.controller.UIController;
import acr.browser.lightning.browser.fragment.anim.HorizontalItemAnimator;
import acr.browser.lightning.browser.fragment.anim.VerticalItemAnimator;
import acr.browser.lightning.preference.PreferenceManager;
import acr.browser.lightning.utils.DrawableUtils;
import acr.browser.lightning.utils.ThemeUtils;
import acr.browser.lightning.utils.Utils;
import acr.browser.lightning.view.BackgroundDrawable;
import acr.browser.lightning.view.LightningView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * A fragment that holds and manages the tabs and interaction with the tabs.
 * It is reliant on the BrowserController in order to get the current UI state
 * of the browser. It also uses the BrowserController to signal that the UI needs
 * to change. This class contains the adapter used by both the drawer tabs and
 * the desktop tabs. It delegates touch events for the tab UI appropriately.
 */
public class TabsFragment extends Fragment implements View.OnClickListener, View.OnLongClickListener, TabsView {

    @NonNull
    public static TabsFragment createTabsFragment(boolean isIncognito, boolean showTabsInDrawer) {
        TabsFragment tabsFragment = new TabsFragment();
        final Bundle tabsFragmentArguments = new Bundle();
        tabsFragmentArguments.putBoolean(TabsFragment.IS_INCOGNITO, isIncognito);
        tabsFragmentArguments.putBoolean(TabsFragment.VERTICAL_MODE, showTabsInDrawer);
        tabsFragment.setArguments(tabsFragmentArguments);

        return tabsFragment;
    }

    private static final String TAG = "TabsFragment";

    /**
     * Arguments boolean to tell the fragment it is displayed in the drawner or on the tab strip
     * If true, the fragment is in the left drawner in the strip otherwise.
     */
    private static final String VERTICAL_MODE = TAG + ".VERTICAL_MODE";
    private static final String IS_INCOGNITO = TAG + ".IS_INCOGNITO";

    private boolean mIsIncognito, mDarkTheme;
    private int mIconColor;
    private boolean mColorMode = true;
    private boolean mShowInNavigationDrawer;

    @Nullable private LightningViewAdapter mTabsAdapter;
    private UIController mUiController;

    @BindView(R.id.tabs_list) RecyclerView mRecyclerView;
    private Unbinder mUnbinder;

    private TabsManager mTabsManager;
    @Inject PreferenceManager mPreferences;

    public TabsFragment() {
        BrowserApp.getAppComponent().inject(this);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final Bundle arguments = getArguments();
        final Context context = getContext();
        mUiController = (UIController) getActivity();
        mTabsManager = mUiController.getTabModel();
        mIsIncognito = arguments.getBoolean(IS_INCOGNITO, false);
        mShowInNavigationDrawer = arguments.getBoolean(VERTICAL_MODE, true);
        mDarkTheme = mPreferences.getUseTheme() != 0 || mIsIncognito;
        mColorMode = mPreferences.getColorModeEnabled();
        mColorMode &= !mDarkTheme;
        mIconColor = mDarkTheme ?
            ThemeUtils.getIconDarkThemeColor(context) :
            ThemeUtils.getIconLightThemeColor(context);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View view;
        final LayoutManager layoutManager;
        if (mShowInNavigationDrawer) {
            view = inflater.inflate(R.layout.tab_drawer, container, false);
            layoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
            setupFrameLayoutButton(view, R.id.tab_header_button, R.id.plusIcon);
            setupFrameLayoutButton(view, R.id.new_tab_button, R.id.icon_plus);
            setupFrameLayoutButton(view, R.id.action_back, R.id.icon_back);
            setupFrameLayoutButton(view, R.id.action_forward, R.id.icon_forward);
            setupFrameLayoutButton(view, R.id.action_home, R.id.icon_home);
        } else {
            view = inflater.inflate(R.layout.tab_strip, container, false);
            layoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false);
            ImageView newTab = view.findViewById(R.id.new_tab_button);
            newTab.setColorFilter(ThemeUtils.getIconDarkThemeColor(getActivity()));
            newTab.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    mUiController.newTabButtonClicked();
                }
            });
        }

        mUnbinder = ButterKnife.bind(this, view);

        SimpleItemAnimator animator;
        if (mShowInNavigationDrawer) {
            animator = new VerticalItemAnimator();
        } else {
            animator = new HorizontalItemAnimator();
        }
        animator.setSupportsChangeAnimations(false);
        animator.setAddDuration(200);
        animator.setChangeDuration(0);
        animator.setRemoveDuration(200);
        animator.setMoveDuration(200);
        mRecyclerView.setLayerType(View.LAYER_TYPE_NONE, null);
        mRecyclerView.setItemAnimator(animator);
        mRecyclerView.setLayoutManager(layoutManager);
        mTabsAdapter = new LightningViewAdapter(mShowInNavigationDrawer);
        mRecyclerView.setAdapter(mTabsAdapter);
        mRecyclerView.setHasFixedSize(true);

        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mUnbinder != null) {
            mUnbinder.unbind();
            mUnbinder = null;
        }
        mTabsAdapter = null;
    }

    private TabsManager getTabsManager() {
        if (mTabsManager == null) {
            mTabsManager = mUiController.getTabModel();
        }
        return mTabsManager;
    }

    private void setupFrameLayoutButton(@NonNull final View root, @IdRes final int buttonId,
                                        @IdRes final int imageId) {
        final View frameButton = root.findViewById(buttonId);
        final ImageView buttonImage = root.findViewById(imageId);
        frameButton.setOnClickListener(this);
        frameButton.setOnLongClickListener(this);
        buttonImage.setColorFilter(mIconColor, PorterDuff.Mode.SRC_IN);
    }

    @Override
    public void onResume() {
        super.onResume();
        // Force adapter refresh
        if (mTabsAdapter != null) {
            mTabsAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void tabsInitialized() {
        if (mTabsAdapter != null) {
            mTabsAdapter.notifyDataSetChanged();
        }
    }

    public void reinitializePreferences() {
        Activity activity = getActivity();
        if (activity == null) {
            return;
        }
        mDarkTheme = mPreferences.getUseTheme() != 0 || mIsIncognito;
        mColorMode = mPreferences.getColorModeEnabled();
        mColorMode &= !mDarkTheme;
        mIconColor = mDarkTheme ?
            ThemeUtils.getIconDarkThemeColor(activity) :
            ThemeUtils.getIconLightThemeColor(activity);
        if (mTabsAdapter != null) {
            mTabsAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void onClick(@NonNull View v) {
        switch (v.getId()) {
            case R.id.tab_header_button:
                mUiController.showCloseDialog(getTabsManager().indexOfCurrentTab());
                break;
            case R.id.new_tab_button:
                mUiController.newTabButtonClicked();
                break;
            case R.id.action_back:
                mUiController.onBackButtonPressed();
                break;
            case R.id.action_forward:
                mUiController.onForwardButtonPressed();
                break;
            case R.id.action_home:
                mUiController.onHomeButtonPressed();
            default:
                break;
        }
    }

    @Override
    public boolean onLongClick(@NonNull View v) {
        switch (v.getId()) {
            case R.id.action_new_tab:
                mUiController.newTabButtonLongClicked();
                break;
            default:
                break;
        }
        return true;
    }

    @Override
    public void tabAdded() {
        if (mTabsAdapter != null) {
            mTabsAdapter.notifyItemInserted(getTabsManager().last());
            mRecyclerView.postDelayed(new Runnable() {
                @Override
                public void run() {
                    mRecyclerView.smoothScrollToPosition(mTabsAdapter.getItemCount() - 1);
                }
            }, 500);
        }
    }

    @Override
    public void tabRemoved(int position) {
        if (mTabsAdapter != null) {
            mTabsAdapter.notifyItemRemoved(position);
        }
    }

    @Override
    public void tabChanged(int position) {
        if (mTabsAdapter != null) {
            mTabsAdapter.notifyItemChanged(position);
        }
    }

    private class LightningViewAdapter extends RecyclerView.Adapter<LightningViewAdapter.LightningViewHolder> {

        private final int mLayoutResourceId;
        @Nullable private final Drawable mBackgroundTabDrawable;
        @Nullable private final Bitmap mForegroundTabBitmap;
        private ColorMatrix mColorMatrix;
        private Paint mPaint;
        private ColorFilter mFilter;
        private static final float DESATURATED = 0.5f;

        private final boolean mDrawerTabs;

        public LightningViewAdapter(final boolean vertical) {
            this.mLayoutResourceId = vertical ? R.layout.tab_list_item : R.layout.tab_list_item_horizontal;
            this.mDrawerTabs = vertical;

            if (vertical) {
                mBackgroundTabDrawable = null;
                mForegroundTabBitmap = null;
            } else {
                int backgroundColor = Utils.mixTwoColors(ThemeUtils.getPrimaryColor(getContext()), Color.BLACK, 0.75f);
                Bitmap backgroundTabBitmap = Bitmap.createBitmap(Utils.dpToPx(175), Utils.dpToPx(30), Bitmap.Config.ARGB_8888);
                Utils.drawTrapezoid(new Canvas(backgroundTabBitmap), backgroundColor, true);
                mBackgroundTabDrawable = new BitmapDrawable(getResources(), backgroundTabBitmap);

                int foregroundColor = ThemeUtils.getPrimaryColor(getContext());
                mForegroundTabBitmap = Bitmap.createBitmap(Utils.dpToPx(175), Utils.dpToPx(30), Bitmap.Config.ARGB_8888);
                Utils.drawTrapezoid(new Canvas(mForegroundTabBitmap), foregroundColor, false);
            }
        }

        @NonNull
        @Override
        public LightningViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
            LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
            View view = inflater.inflate(mLayoutResourceId, viewGroup, false);
            if (mDrawerTabs) {
                DrawableUtils.setBackground(view, new BackgroundDrawable(view.getContext()));
            }
            return new LightningViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull final LightningViewHolder holder, int position) {
            holder.exitButton.setTag(position);

            ViewCompat.jumpDrawablesToCurrentState(holder.exitButton);

            LightningView web = getTabsManager().getTabAtPosition(position);
            if (web == null) {
                return;
            }
            holder.txtTitle.setText(web.getTitle());

            final Bitmap favicon = web.getFavicon();
            if (web.isForegroundTab()) {
                Drawable foregroundDrawable = null;
                if (!mDrawerTabs) {
                    foregroundDrawable = new BitmapDrawable(getResources(), mForegroundTabBitmap);
                    if (!mIsIncognito && mColorMode) {
                        foregroundDrawable.setColorFilter(mUiController.getUiColor(), PorterDuff.Mode.SRC_IN);
                    }
                }
                if (!mIsIncognito && mColorMode) {
                    mUiController.changeToolbarBackground(favicon, foregroundDrawable);
                }

                TextViewCompat.setTextAppearance(holder.txtTitle, R.style.boldText);
                if (!mDrawerTabs) {
                    DrawableUtils.setBackground(holder.layout, foregroundDrawable);
                }
                holder.favicon.setImageBitmap(favicon);
            } else {
                TextViewCompat.setTextAppearance(holder.txtTitle, R.style.normalText);
                if (!mDrawerTabs) {
                    DrawableUtils.setBackground(holder.layout, mBackgroundTabDrawable);
                }
                holder.favicon.setImageBitmap(getDesaturatedBitmap(favicon));
            }

            if (mDrawerTabs) {
                BackgroundDrawable verticalBackground = (BackgroundDrawable) holder.layout.getBackground();
                verticalBackground.setCrossFadeEnabled(false);
                if (web.isForegroundTab()) {
                    verticalBackground.startTransition(200);
                } else {
                    verticalBackground.reverseTransition(200);
                }
            }
        }

        @Override
        public int getItemCount() {
            return getTabsManager().size();
        }

        public Bitmap getDesaturatedBitmap(@NonNull Bitmap favicon) {
            Bitmap grayscaleBitmap = Bitmap.createBitmap(favicon.getWidth(),
                favicon.getHeight(), Bitmap.Config.ARGB_8888);

            Canvas c = new Canvas(grayscaleBitmap);
            if (mColorMatrix == null || mFilter == null || mPaint == null) {
                mPaint = new Paint();
                mColorMatrix = new ColorMatrix();
                mColorMatrix.setSaturation(DESATURATED);
                mFilter = new ColorMatrixColorFilter(mColorMatrix);
                mPaint.setColorFilter(mFilter);
            }

            c.drawBitmap(favicon, 0, 0, mPaint);
            return grayscaleBitmap;
        }

        public class LightningViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {

            public LightningViewHolder(@NonNull View view) {
                super(view);
                txtTitle = view.findViewById(R.id.textTab);
                favicon = view.findViewById(R.id.faviconTab);
                exit = view.findViewById(R.id.deleteButton);
                layout = view.findViewById(R.id.tab_item_background);
                exitButton = view.findViewById(R.id.deleteAction);
                exit.setColorFilter(mIconColor, PorterDuff.Mode.SRC_IN);

                exitButton.setOnClickListener(this);
                layout.setOnClickListener(this);
                layout.setOnLongClickListener(this);
            }

            @NonNull final TextView txtTitle;
            @NonNull final ImageView favicon;
            @NonNull final ImageView exit;
            @NonNull final FrameLayout exitButton;
            @NonNull final LinearLayout layout;

            @Override
            public void onClick(View v) {
                if (v == exitButton) {
                    mUiController.tabCloseClicked(getAdapterPosition());
                }
                if (v == layout) {
                    mUiController.tabClicked(getAdapterPosition());
                }
            }

            @Override
            public boolean onLongClick(View v) {
                mUiController.showCloseDialog(getAdapterPosition());
                return true;
            }
        }
    }
}

File: app/src/main/java/acr/browser/lightning/browser/fragment/TabsFragment.kt
package acr.browser.lightning.browser.fragment

import acr.browser.lightning.R
import acr.browser.lightning.browser.TabsManager
import acr.browser.lightning.browser.TabsView
import acr.browser.lightning.browser.fragment.anim.HorizontalItemAnimator
import acr.browser.lightning.browser.fragment.anim.VerticalItemAnimator
import acr.browser.lightning.controller.UIController
import acr.browser.lightning.di.injector
import acr.browser.lightning.extensions.color
import acr.browser.lightning.extensions.desaturate
import acr.browser.lightning.extensions.drawTrapezoid
import acr.browser.lightning.preference.UserPreferences
import acr.browser.lightning.utils.DrawableUtils
import acr.browser.lightning.utils.ThemeUtils
import acr.browser.lightning.utils.Utils
import acr.browser.lightning.view.BackgroundDrawable
import acr.browser.lightning.view.LightningView
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.IdRes
import androidx.core.widget.TextViewCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.tab_drawer.*
import java.util.*
import javax.inject.Inject

/**
 * A fragment that holds and manages the tabs and interaction with the tabs. It is reliant on the
 * BrowserController in order to get the current UI state of the browser. It also uses the
 * BrowserController to signal that the  the desktop tabs. It delegates touch events for the tab UI
 * appropriately.
 */
class TabsFragment : Fragment(), View.OnClickListener, View.OnLongClickListener, TabsView {

    private var isIncognito: Boolean = false
    private var darkTheme: Boolean = false
    private var iconColor: Int = 0
    private var colorMode = true
    private var showInNavigationDrawer: Boolean = false

    private var tabsAdapter: LightningViewAdapter? = null
    private lateinit var uiController: UIController

    @Inject internal lateinit var userPreferences: UserPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        injector.inject(this)
        val context = requireNotNull(context) { "Context should never be null in onCreate" }
        uiController = activity as UIController
        isIncognito = arguments?.getBoolean(IS_INCOGNITO, false) == true
        showInNavigationDrawer = arguments?.getBoolean(VERTICAL_MODE, true) == true
        darkTheme = userPreferences.useTheme != 0 || isIncognito
        colorMode = userPreferences.colorModeEnabled
        colorMode = colorMode and !darkTheme

        iconColor = ThemeUtils.getIconThemeColor(context, darkTheme)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view: View
        val context = inflater.context
        if (showInNavigationDrawer) {
            view = inflater.inflate(R.layout.tab_drawer, container, false)
            setupFrameLayoutButton(view, R.id.tab_header_button, R.id.plusIcon)
            setupFrameLayoutButton(view, R.id.new_tab_button, R.id.icon_plus)
            setupFrameLayoutButton(view, R.id.action_back, R.id.icon_back)
            setupFrameLayoutButton(view, R.id.action_forward, R.id.icon_forward)
            setupFrameLayoutButton(view, R.id.action_home, R.id.icon_home)
        } else {
            view = inflater.inflate(R.layout.tab_strip, container, false)
            view.findViewById<ImageView>(R.id.new_tab_button).apply {
                setColorFilter(context.color(R.color.icon_dark_theme))
                setOnClickListener(this@TabsFragment)
                setOnLongClickListener(this@TabsFragment)
            }
        }

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val layoutManager = if (showInNavigationDrawer) {
            LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        } else {
            LinearLayoutManager(context, RecyclerView.HORIZONTAL, false)
        }

        val animator = (if (showInNavigationDrawer) {
            VerticalItemAnimator()
        } else {
            HorizontalItemAnimator()
        }).apply {
            supportsChangeAnimations = false
            addDuration = 200
            changeDuration = 0
            removeDuration = 200
            moveDuration = 200
        }

        tabsAdapter = LightningViewAdapter(showInNavigationDrawer)

        tabs_list.apply {
            setLayerType(View.LAYER_TYPE_NONE, null)
            itemAnimator = animator
            this.layoutManager = layoutManager
            adapter = tabsAdapter
            setHasFixedSize(true)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        tabsAdapter = null
    }

    private fun getTabsManager(): TabsManager = uiController.getTabModel()

    private fun setupFrameLayoutButton(root: View, @IdRes buttonId: Int,
                                       @IdRes imageId: Int) {
        val frameButton = root.findViewById<View>(buttonId)
        val buttonImage = root.findViewById<ImageView>(imageId)
        frameButton.setOnClickListener(this)
        frameButton.setOnLongClickListener(this)
        buttonImage.setColorFilter(iconColor, PorterDuff.Mode.SRC_IN)
    }

    override fun onResume() {
        super.onResume()
        // Force adapter refresh
        tabsAdapter?.notifyDataSetChanged()
    }

    override fun tabsInitialized() {
        tabsAdapter?.notifyDataSetChanged()
    }

    fun reinitializePreferences() {
        val activity = activity ?: return
        darkTheme = userPreferences.useTheme != 0 || isIncognito
        colorMode = userPreferences.colorModeEnabled
        colorMode = colorMode and !darkTheme
        iconColor = ThemeUtils.getIconThemeColor(activity, darkTheme)
        tabsAdapter?.notifyDataSetChanged()
    }

    override fun onClick(v: View) = when (v.id) {
        R.id.tab_header_button -> uiController.showCloseDialog(getTabsManager().indexOfCurrentTab())
        R.id.new_tab_button -> uiController.newTabButtonClicked()
        R.id.action_back -> uiController.onBackButtonPressed()
        R.id.action_forward -> uiController.onForwardButtonPressed()
        R.id.action_home -> uiController.onHomeButtonPressed()
        else -> {
        }
    }

    override fun onLongClick(v: View): Boolean {
        when (v.id) {
            R.id.new_tab_button -> uiController.newTabButtonLongClicked()
            else -> {
            }
        }
        return true
    }

    override fun tabAdded() {
        tabsAdapter?.let {
            it.showTabs(toViewModels(getTabsManager().allTabs))
            tabs_list.postDelayed({ tabs_list.smoothScrollToPosition(it.itemCount - 1) }, 500)
        }
    }

    override fun tabRemoved(position: Int) {
        tabsAdapter?.showTabs(toViewModels(getTabsManager().allTabs))
    }

    override fun tabChanged(position: Int) {
        tabsAdapter?.showTabs(toViewModels(getTabsManager().allTabs))
    }

    private fun toViewModels(tabs: List<LightningView>) = tabs.map(::TabViewState)

    private inner class LightningViewAdapter(
        private val drawerTabs: Boolean
    ) : RecyclerView.Adapter<LightningViewAdapter.LightningViewHolder>() {

        private val layoutResourceId: Int = if (drawerTabs) R.layout.tab_list_item else R.layout.tab_list_item_horizontal
        private val backgroundTabDrawable: Drawable?
        private val foregroundTabBitmap: Bitmap?
        private var tabList: List<TabViewState> = ArrayList()

        init {

            if (drawerTabs) {
                backgroundTabDrawable = null
                foregroundTabBitmap = null
            } else {
                val context = requireNotNull(context) { "Adapter cannot be initialized when fragment is detached" }

                val backgroundColor = Utils.mixTwoColors(ThemeUtils.getPrimaryColor(context), Color.BLACK, 0.75f)
                val backgroundTabBitmap = Bitmap.createBitmap(Utils.dpToPx(175f), Utils.dpToPx(30f), Bitmap.Config.ARGB_8888).also {
                    Canvas(it).drawTrapezoid(backgroundColor, true)
                }
                backgroundTabDrawable = BitmapDrawable(resources, backgroundTabBitmap)

                val foregroundColor = ThemeUtils.getPrimaryColor(context)
                foregroundTabBitmap = Bitmap.createBitmap(Utils.dpToPx(175f), Utils.dpToPx(30f), Bitmap.Config.ARGB_8888).also {
                    Canvas(it).drawTrapezoid(foregroundColor, false)
                }
            }
        }

        internal fun showTabs(tabs: List<TabViewState>) {
            val oldList = tabList
            tabList = ArrayList(tabs)

            val result = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize() = oldList.size

                override fun getNewListSize() = tabList.size

                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) =
                    oldList[oldItemPosition] == tabList[newItemPosition]

                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    val oldTab = oldList[oldItemPosition]
                    val newTab = tabList[newItemPosition]

                    return (oldTab.title == newTab.title
                        && oldTab.favicon == newTab.favicon
                        && oldTab.isForegroundTab == newTab.isForegroundTab
                        && oldTab == newTab)
                }
            })

            result.dispatchUpdatesTo(this)
        }

        override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): LightningViewHolder {
            val inflater = LayoutInflater.from(viewGroup.context)
            val view = inflater.inflate(layoutResourceId, viewGroup, false)
            if (drawerTabs) {
                DrawableUtils.setBackground(view, BackgroundDrawable(view.context))
            }
            return LightningViewHolder(view)
        }

        override fun onBindViewHolder(holder: LightningViewHolder, position: Int) {
            holder.exitButton.tag = position

            holder.exitButton.jumpDrawablesToCurrentState()

            val web = tabList[position]

            updateViewHolderTitle(holder, web.title)
            updateViewHolderAppearance(holder, web.favicon, web.isForegroundTab)
            updateViewHolderFavicon(holder, web.favicon, web.isForegroundTab)
            updateViewHolderBackground(holder, web.isForegroundTab)
        }

        private fun updateViewHolderTitle(viewHolder: LightningViewHolder, title: String) {
            viewHolder.txtTitle.text = title
        }

        private fun updateViewHolderFavicon(viewHolder: LightningViewHolder, favicon: Bitmap, isForeground: Boolean) =
            if (isForeground) {
                viewHolder.favicon.setImageBitmap(favicon)
            } else {
                viewHolder.favicon.setImageBitmap(favicon.desaturate())
            }

        private fun updateViewHolderBackground(viewHolder: LightningViewHolder, isForeground: Boolean) {
            if (drawerTabs) {
                val verticalBackground = viewHolder.layout.background as BackgroundDrawable
                verticalBackground.isCrossFadeEnabled = false
                if (isForeground) {
                    verticalBackground.startTransition(200)
                } else {
                    verticalBackground.reverseTransition(200)
                }
            }
        }

        private fun updateViewHolderAppearance(viewHolder: LightningViewHolder, favicon: Bitmap, isForeground: Boolean) {
            if (isForeground) {
                var foregroundDrawable: Drawable? = null
                if (!drawerTabs) {
                    foregroundDrawable = BitmapDrawable(resources, foregroundTabBitmap)
                    if (!isIncognito && colorMode) {
                        foregroundDrawable.setColorFilter(uiController.getUiColor(), PorterDuff.Mode.SRC_IN)
                    }
                }
                TextViewCompat.setTextAppearance(viewHolder.txtTitle, R.style.boldText)
                if (!drawerTabs) {
                    DrawableUtils.setBackground(viewHolder.layout, foregroundDrawable)
                }
                if (!isIncognito && colorMode) {
                    uiController.changeToolbarBackground(favicon, foregroundDrawable)
                }
            } else {
                TextViewCompat.setTextAppearance(viewHolder.txtTitle, R.style.normalText)
                if (!drawerTabs) {
                    DrawableUtils.setBackground(viewHolder.layout, backgroundTabDrawable)
                }
            }
        }

        override fun getItemCount() = tabList.size

        internal inner class LightningViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener, View.OnLongClickListener {

            val txtTitle: TextView = view.findViewById(R.id.textTab)
            val favicon: ImageView = view.findViewById(R.id.faviconTab)
            val exit: ImageView = view.findViewById(R.id.deleteButton)
            val exitButton: FrameLayout = view.findViewById(R.id.deleteAction)
            val layout: LinearLayout = view.findViewById(R.id.tab_item_background)

            init {
                exit.setColorFilter(iconColor, PorterDuff.Mode.SRC_IN)
                exitButton.setOnClickListener(this)
                layout.setOnClickListener(this)
                layout.setOnLongClickListener(this)
            }

            override fun onClick(v: View) {
                if (v === exitButton) {
                    uiController.tabCloseClicked(adapterPosition)
                } else if (v === layout) {
                    uiController.tabClicked(adapterPosition)
                }
            }

            override fun onLongClick(v: View): Boolean {
                uiController.showCloseDialog(adapterPosition)
                return true
            }
        }

    }

    companion object {

        @JvmStatic
        fun createTabsFragment(isIncognito: Boolean, showTabsInDrawer: Boolean): TabsFragment {
            return TabsFragment().apply {
                arguments = Bundle().apply {
                    putBoolean(TabsFragment.IS_INCOGNITO, isIncognito)
                    putBoolean(TabsFragment.VERTICAL_MODE, showTabsInDrawer)
                }
            }
        }

        private const val TAG = "TabsFragment"

        private const val DESATURATED = 0.5f

        /**
         * Arguments boolean to tell the fragment it is displayed in the drawner or on the tab strip
         * If true, the fragment is in the left drawner in the strip otherwise.
         */
        private const val VERTICAL_MODE = "$TAG.VERTICAL_MODE"
        private const val IS_INCOGNITO = "$TAG.IS_INCOGNITO"
    }
}
